Firrtl instance flips

mikeurbach opened the pull request:
This is to address #261. I separated the change into: a) make LowerTypes optional into firtool, b) add more verification for instance result types, and c) fix LowerTypes to handle the situation brought up in the issue. I hope we can merge all three changes, but I separated them in case we want to cherry pick.

mikeurbach commented on test/firrtl/errors.mlir:
```diff
@@ -290,4 +290,43 @@ firrtl.module @X(%a : !firrtl.uint<4>) {
+firrtl.circuit "X" {
+
+firrtl.module @X(%a : !firrtl.uint<1>, %b : !firrtl.flip<sint<1>>) {
+  // expected-error @+1 {{type mismatch between destination '!firrtl.flip<sint<1>>' and source '!firrtl.uint<1>'}}
```
crap, didn't mean to include this. was working on this file in another branch.

lattner commented on tools/firtool/firtool.cpp:
```diff
@@ -54,6 +54,11 @@ static cl::opt<bool> disableOptimization("disable-opt",
+static cl::opt<bool>
+    enableLowerTypes("enable-lower-types",
+                     cl::desc("run the lower-types pass within lower-to-rtl"),
+                     cl::init(false));
```
Please add a testcase that shows an example working with both modes (one that works, one that produces an error)

mikeurbach replied:
@lattner is this what you had in mind? https://github.com/llvm/circt/pull/263/commits/0a64166528cf0b6bc82be3ac0089d1a4b531cdb9

In this PR, I fixed the issue Andrew brought up that led me to want to disable this by default. But this demonstrates the main issue that pass was created for--bundle types cannot be lowered to RTL without it.

lattner replied:
Looks great, thank you Mike!

lattner reviewed (approved):
Thanks!

darthscsi commented:
Seems correct on my tests.

darthscsi reviewed (approved):
LGTM

mikeurbach commented on lib/Dialect/FIRRTL/LowerTypes.cpp:
```diff
@@ -276,8 +295,13 @@ void FIRRTLTypesLowering::visitStmt(ConnectOp op) {
+
   // If we aren't connecting two bundles, there is nothing to do.
-  if (!dest.getType().isa<BundleType>() || !src.getType().isa<BundleType>())
+  if (!destType || !srcType)
```
@lattner I think I finally understand your question here about outer level flip types: https://github.com/llvm/circt/pull/157#discussion_r519233669. I was getting bit here trying to write this [test case](https://github.com/llvm/circt/pull/263/files#diff-cfcb11e78479b63cf6e909ee4a9e01a0ff1ba2ab019da32a0c64baa56b7e0c12R143). I've pulled the logic to peel off the outer flip so I could use it here as well.

mikeurbach commented:
I found and fixed another place where LowerTypes didn't handle a bundle type wrapped in an outer level flip type, and added a test case for the new scenario. I also added test cases showing what happens with and without `-enable-lower-types` when using `firtool` with modules that have bundle typed ports.

I'd like to merge this so as to not block anyone testing FIRRTL features, so I'm going to request some post-commit review on these last two items.

darthscsi commented:
Thanks for looking at it so quickly!

On Fri, Nov 20, 2020 at 1:55 PM mikeurbach <notifications@github.com> wrote:

> Merged #263 <https://github.com/llvm/circt/pull/263> into master.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/circt/pull/263#event-4022541870>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACY5KKYKCIMVNU67HZDD77TSQ3CSVANCNFSM4T3Z46UQ>
> .
>
